realtek: Fix MMD-register access on RTL838x

oliv3r opened the pull request:
This fixes 2 bugs in MMD access on the RTL838x:
	- We need to preserve the state of the park-page because the normal PHY r/w functions expect a correct setting in RTL838X_SMI_ACCESS_PHY_CTRL_1
	- commit b53202a8c3f728c348c5376e5b5fb36af7c37744 introduced calling the hw MMD access functions with mdiobus_c45_devad(regnum) so now we need to invert this op via mdiobus_c45_regad(reg)

This was tested with MMD access register R/W for EEE functionality on the Zyxel GS1900-10HP.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Preserve park-page register content

We need to preserve the state of the park-page because the normal PHY
r/w functions expect a correct setting in RTL838X_SMI_ACCESS_PHY_CTRL_1.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>

Commit:
realtek: Fix MMD-register access on RTL838x

In commit b53202a8c3f7 ("realtek: switch to use generic MDIO accessor
functions") introduced calling the hw MMD access functions with
mdiobus_c45_devad(regnum) so lets follow suit.

Signed-off-by: Birger Koblitz <git@birger-koblitz.de>
Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r commented:
@svanheule a patch in a series of 'small fixes' as individual MR's.

svanheule commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
-	sw_w32(v, RTL838X_SMI_ACCESS_PHY_CTRL_1);
+	park_page = sw_r32(RTL838X_SMI_ACCESS_PHY_CTRL_1) & ((0x1f << 15) | 0x2);
+	/* mmd-access | read: BIT(2) not set | cmd-start */
+	v = BIT(1) | BIT(0);
```
And TYPE_MMD is again set here. That's a bit redundant.

oliv3r replied:
That I can clean up :)

svanheule commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
-	/* mmd-access | read | cmd-start */
-	v = 1 << 1 | 0 << 2 | 1;
-	sw_w32(v, RTL838X_SMI_ACCESS_PHY_CTRL_1);
+	park_page = sw_r32(RTL838X_SMI_ACCESS_PHY_CTRL_1) & ((0x1f << 15) | 0x2);
```
This extracts the PARK_PAGE field _and_ hard sets the TYPE field (to MMD). AFAICT PARK_PAGE isn't actually set anywhere at all, so does this just depend on the last value it was set to before this driver runs?

Do you know what value is stored here? According to the SDK, a value of 0x1f means something along the lines of "ignore parked page". I have no idea what a "parked page" is however, but even the SDK seems to almost always use 0x1f. I've found a few occurrences of `PARK_PAGE = 0`

oliv3r replied:
the PARK_PAGE MR is now #11090 and we can move the discussion there.

svanheule commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
 	if (rtl838x_smi_wait_op(100000))
 		goto timeout;
 
+	sw_w32(park_page, RTL838X_SMI_ACCESS_PHY_CTRL_1);
```
Does it really matter if you leave the last only PARK_PAGE (and operation type) set in this register? Seems a bit fragile to depend on keeping state in this type of register.

oliv3r replied:
I whole-heatedly agree with you here. And I admit you've looked into this patch in far more detail then I have!

But you are right, there's a lot of assumptions happening in this entire driver ...

svanheule reviewed (commented):
I feel like this warrants two separate commits (Fixes: tags etc)

oliv3r commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -1891,27 +1891,27 @@ int rtl838x_write_phy(u32 port, u32 page, u32 reg, u32 val)
-	if (rtl838x_smi_wait_op(100000))
-		goto timeout;
-
-	sw_w32(1 << port, RTL838X_SMI_ACCESS_PHY_CTRL_0);
```
You do realize I can't answer you this question :D

oliv3r commented:
So in 'v2', I kept things more conservative, and used 'logic' (which we know realtek is not), so I really need some help in figuring out (and testing) if this is all right. I'm not sure how this was tested before either.

oliv3r committed:
realtek: Define RTL838x SMI Ctrl registers

We currently use a lot of magic values to read/write the SMI registers.
Lets define the ones we know, so we have less magic to deal with.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
realtek: Use SMI ctrl defines instead of magic

With the previously introduced defines, we can remove some magic values!

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>
